core/src/main/scala/org/broadinstitute/dsde/rawls/entities/local/LocalEntityProvider.scala

```diff
@@ -29,9 +31,34 @@ class LocalEntityProvider(workspace: Workspace, implicit protected val dataSourc
 
   private val workspaceContext = workspace
 
-  override def entityTypeMetadata(): Future[Map[String, EntityTypeMetadata]] = {
+  override def entityTypeMetadata(useCache: Boolean): Future[Map[String, EntityTypeMetadata]] = {
     dataSource.inTransaction { dataAccess =>
-      dataAccess.entityQuery.getEntityTypeMetadata(workspaceContext)
+      dataAccess.workspaceQuery.isEntityCacheCurrent(workspaceContext.workspaceIdAsUUID).flatMap { isCacheCurrentOpt =>
+        //If the cache doesn't exist; return full query results and mark the workspace for an update
+        if(isCacheCurrentOpt.isEmpty) {
+          dataAccess.workspaceQuery.updateCacheLastUpdated(workspaceContext.workspaceIdAsUUID, new Timestamp(workspaceContext.lastModified.getMillis - 1)).flatMap { _ =>
+            dataAccess.entityQuery.getEntityTypeMetadata(workspaceContext)
+          }
+        }
```

Comment:
1. I'm wondering what happens here, in the case where `getEntityTypeMetadata` is too slow ... is there a case in which `entityQuery.getEntityTypeMetadata` fails, thus the transaction gets rolled back, thus `updateCacheLastUpdated` is also rolled back and we don't mark this workspace as needing its cache calculated?

2. If the cache exists, but is out of date ... the call to `updateCacheLastUpdated` will change the cache-last-updated value, obscuring any calculation of how out of date it was previously.

Reply:
1. That's a good point, there is probably room for improvement for error-handling here. I will put some more thought into this.

2. That shouldn't happen. If `isCacheCurrentOpt.isEmpty` = `false` and `isCacheCurrentOpt.getOrElse(false)` = `false`, then it will fall to the `else` branch at the end which does not update the timestamp. `isCacheCurrentOpt` is an `Option[Boolean]` that has three modes that dictate: cache is up to date, cache is not up to date, cache does not exist. Looking at it now, the readability isn't the greatest. I'll see if there's a better way to represent these states. At the very least I will add a test for this.

Reply:
code here looks correct. For readability, it would be great if `cacheStatus` was an enum; I wouldn't hold the PR just for that.